test/common/json/json_sanitizer_fuzz_test.cc

```diff
@@ -0,0 +1,35 @@
+#include "source/common/json/json_sanitizer.h"
+#include "source/common/protobuf/utility.h"
+
+#include "test/common/json/json_sanitizer_test_util.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/fuzz/utility.h"
+#include "test/test_common/utility.h"
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Fuzz {
+
+DEFINE_FUZZER(const uint8_t* buf, size_t len) {
+  FuzzedDataProvider provider(buf, len);
+  std::string buffer1, buffer2, errmsg;
+  while (provider.remaining_bytes() != 0) {
+    std::string input = provider.ConsumeRandomLengthString(provider.remaining_bytes());
+    absl::string_view sanitized = Envoy::Json::sanitize(buffer1, input);
+
+    // If the input is valid UTF-8 we can do a differential test against the
+    // Protobuf JSON sanitizer. Otherwise we are simply ensuring that the
+    // sanitizer does not crash.
+    if (Envoy::Json::TestUtil::isProtoSerializableUtf8(input)) {
```

Comment:
Do you have any stats on how frequently we get serializable vs. non-serializable? Asking since this is a good litmus test on the `isProtoSErializableUtf8` itself.

Reply:
Not sure. Worth just a quick manual validation with logs to start with.

Reply:
ok i'll try to do that. I seem to not be able to build with asan-fuzzer consistently; something in my build flow is flaky or non-deterministc. I was able to build it and run it for hours a few days ago.

Reply:
I added some logging so every 10k iterations I print the #serializable/total. It starts of with 2415/10k serializable (20%) but I think the fuzz system learns that the non-serializable paths hit less code, and aby the time I get to 1M entries it's 725k/1M, or 75% serializable.

Not sure if that's what you were looking for but I think it's working as expected.

Reply:
here's the code diff to print out the serializability ratio:
```
diff --git a/test/common/json/json_sanitizer_fuzz_test.cc b/test/common/json/json_sanitizer_fuzz_test.cc
index 83c49fd830..48a9288ed7 100644
--- a/test/common/json/json_sanitizer_fuzz_test.cc
+++ b/test/common/json/json_sanitizer_fuzz_test.cc
@@ -1,3 +1,5 @@
+#include <ostream>
+
 #include "source/common/json/json_sanitizer.h"
 #include "source/common/protobuf/utility.h"
 
@@ -14,6 +16,8 @@ namespace Fuzz {
 DEFINE_FUZZER(const uint8_t* buf, size_t len) {
   FuzzedDataProvider provider(buf, len);
   std::string buffer1, buffer2, errmsg;
+  static uint64_t num_tests = 0;
+  static uint64_t num_serializable_tests = 0;
   while (provider.remaining_bytes() != 0) {
     std::string input = provider.ConsumeRandomLengthString(provider.remaining_bytes());
     absl::string_view sanitized = Envoy::Json::sanitize(buffer1, input);
@@ -21,13 +25,19 @@ DEFINE_FUZZER(const uint8_t* buf, size_t len) {
     // If the input is valid UTF-8 we can do a differential test against the
     // Protobuf JSON sanitizer. Otherwise we are simply ensuring that the
     // sanitizer does not crash.
+    ++num_tests;
     if (Envoy::Json::TestUtil::isProtoSerializableUtf8(input)) {
+      ++num_serializable_tests;
       buffer2 =
           MessageUtil::getJsonStringFromMessageOrDie(ValueUtil::stringValue(input), false, true);
       absl::string_view proto_sanitized = Envoy::Json::stripDoubleQuotes(buffer2);
       RELEASE_ASSERT(Envoy::Json::TestUtil::utf8Equivalent(sanitized, proto_sanitized, errmsg),
                      errmsg);
     }
+    if ((num_tests % 10000) == 0) {
+      std::cerr << "### " << num_serializable_tests << "/" << num_tests << " are serializable"
+                << std::endl;
+    }
   }
 }
```